pkg/scheduler/core/extender.go

```diff
@@ -94,10 +96,112 @@ func NewHTTPExtender(config *schedulerapi.ExtenderConfig) (algorithm.SchedulerEx
 	}, nil
 }
 
+// SupportsPreemption returns if a extender support preemption.
+// A extender should have preempt verb defined and enabled its own node cache.
+func (h *HTTPExtender) SupportsPreemption() bool {
+	return len(h.preemptVerb) > 0
+}
+
+// ProcessPreemption returns filtered candidate nodes and victims after running preemption logic in extender.
+func (h *HTTPExtender) ProcessPreemption(
+	pod *v1.Pod,
+	nodeToVictims map[*v1.Node]*schedulerapi.Victims,
+	nodeNameToInfo map[string]*schedulercache.NodeInfo,
+) (map[*v1.Node]*schedulerapi.Victims, error) {
+	var (
+		result schedulerapi.ExtenderPreemptionResult
+		args   *schedulerapi.ExtenderPreemptionArgs
+	)
+
+	if len(h.preemptVerb) == 0 {
+		return nil, fmt.Errorf("preempt verb is not defined for extender %v but run into ProcessPreemption", h.extenderURL)
+	}
+
+	restoreMap := map[*schedulerapi.MetaPod]*v1.Pod{}
+
+	if h.nodeCacheCapable {
+		// If extender has cached node info, pass NodeNameToMetaVictims in args.
+		nodeNameToVictims := convertToNodeNameToMetaVictims(nodeToVictims, restoreMap)
+		args = &schedulerapi.ExtenderPreemptionArgs{
+			Pod: pod,
+			NodeNameToMetaVictims: nodeNameToVictims,
+		}
+	} else {
+		args = &schedulerapi.ExtenderPreemptionArgs{
+			Pod:           pod,
+			NodeToVictims: nodeToVictims,
+		}
+	}
+
+	if err := h.send(h.preemptVerb, args, &result); err != nil {
+		return nil, err
+	}
+
+	if h.nodeCacheCapable {
+		if result.NodeNameToMetaVictims == nil {
+			return nil, fmt.Errorf(
+				"Extender: %v returns nil `NodeNameToMetaVictims` in result but it claims to have cached node info",
+				h.extenderURL)
+		}
+		// If extender has cached node info, restore NodeToVictims from NodeNameToInfo and return.
+		nodeToVictims := convertToNodeToVictims(result.NodeNameToMetaVictims, restoreMap, nodeNameToInfo)
+		return nodeToVictims, nil
+	} else {
+		return result.NodeToVictims, nil
+	}
+}
+
+// convertToNodeToVictims converts from meta types to struct type.
+func convertToNodeToVictims(
+	nodeNameToMetaVictims map[string]*schedulerapi.MetaVictims,
+	restoreMap map[*schedulerapi.MetaPod]*v1.Pod,
+	nodeNameToInfo map[string]*schedulercache.NodeInfo,
+) map[*v1.Node]*schedulerapi.Victims {
+	nodeToVictims := map[*v1.Node]*schedulerapi.Victims{}
+	for nodeName, metaVictims := range nodeNameToMetaVictims {
+		victims := &schedulerapi.Victims{
+			Pods: []*v1.Pod{},
+		}
+		for _, metaPod := range metaVictims.Pods {
+			pod := restoreMap[metaPod]
```

Comment:
We shouldn't assume that extenders never add new victims. As a result, `metaPod` may not exist in `restoreMap`.